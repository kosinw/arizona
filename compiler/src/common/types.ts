import { Syn, SyntaxType } from "../parser/types";

export type TraverseMapFunctor = (node: Syn, walk?: TraverseMapFunctor) => void; 

export type TraverseMap = {
    [x in SyntaxType | '*']?: TraverseMapFunctor;
};