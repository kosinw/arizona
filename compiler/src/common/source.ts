import invariant from 'invariant';
import lineColumn from 'line-column';

export interface LineColumn {
  line: number;
  column: number;
}

export interface LineColumnSpan {
  start: LineColumn;
  end: LineColumn;
}

export class SourceSpan {
  constructor(
    public start: number,
    public end: number,
    public source: SourceFile
  ) {
    invariant(
      start >= 0,
      'start of span cannot be earlier than first character'
    );
    invariant(start < end, 'start of span must come before end');
    invariant(
      end <= source.length,
      'end of span cannot be past source file length'
    );
  }

  get length() {
    return this.end - this.start;
  }

  getLineColumnFromIndex(idx: number): LineColumn {
    const { line, col } = lineColumn(this.source.contents).fromIndex(idx)!;

    return { line, column: col };
  }

  get cursor(): LineColumnSpan {
    return {
      start: this.getLineColumnFromIndex(this.start),
      end: this.getLineColumnFromIndex(this.end),
    };
  }

  combine(other: SourceSpan): SourceSpan {
    invariant(
      this.source === other.source,
      'both spans must have to same original source file'
    );

    const start = Math.min(this.start, other.start);
    const end = Math.max(this.end, other.end);

    return new SourceSpan(start, end, this.source);
  }

  static join(...spans: SourceSpan[]): SourceSpan {
    let combined = spans.pop();

    invariant(
      !!combined,
      'at least first span provided must not be null/undefined'
    );

    for (let span of spans) {
      combined = combined?.combine(span);
    }

    return combined;
  }
}

export class SourceFile {
  constructor(public contents: string, public filepath: string) {}

  span(start: number, end: number) {
    return new SourceSpan(start, end, this);
  }

  get length() {
    return this.contents.length;
  }
}
