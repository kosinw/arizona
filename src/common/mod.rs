pub mod source;
pub mod span;