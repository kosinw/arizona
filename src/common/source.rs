use std::{
    path::PathBuf,
    io::Read,
    fs::File
};

#[derive(Debug, PartialEq, Clone)]
pub struct Source {
    pub contents: String,
    pub path: PathBuf
}

impl Source {
    pub fn new(contents: &str, path: &PathBuf) -> Source {
        Source {
            contents: contents.to_string(),
            path: path.clone()
        }
    }

    pub fn path(path: PathBuf) -> std::io::Result<Source> {
        let mut contents = String::new();
        let mut file = File::open(path.clone())?;
        file.read_to_string(&mut contents)?;

        Ok(Source::new(&contents, &path))
    }

    pub fn pathless(source: &str) -> Source {
        Source::new(&source.to_string(), &PathBuf::from("[no path]"))
    }
}