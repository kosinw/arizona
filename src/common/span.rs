use crate::common::source::Source;
use std::rc::Rc;

#[derive(Debug, Clone, PartialEq)]
pub struct Span {
    pub offset: usize,
    pub length: usize,
    pub source: Option<Rc<Source>> 
}

#[derive(Debug, Clone, PartialEq)]
pub struct Spanned<T> {
    pub item: T,
    pub span: Span,
}

impl Span {
    pub fn new(source: &Rc<Source>, offset: usize, length: usize) -> Span {
        Span { source: Some(Rc::clone(source)), offset, length }
    }

    pub fn point(source: &Rc<Source>, offset: usize) -> Span {
        // NOTE: maybe it should be 0?
        Span { source: Some(Rc::clone(source)), offset, length: 0 }
    }

    pub fn empty() -> Span {
        Span { source: None, offset: 0, length: usize::MAX }
    }

    pub fn is_empty(&self) -> bool {
        self.source == None
    }

    pub fn end(&self) -> usize {
        self.offset + self.length
    }

    pub fn later_than(&self, other: &Span) -> bool {
        self.offset > other.offset
           || (self.offset == other.offset
              && self.end() > other.end())
    }

    pub fn combine(a: &Span, b: &Span) -> Span {
        if a.is_empty() { return b.clone(); }
        if b.is_empty() { return a.clone(); }

        if a.source != b.source {
            panic!("Can't combine two Spans with separate sources")
        }

        let offset = a.offset.min(b.offset);
        let end    = a.end().max(b.end());
        let length = end - offset;

        return Span::new(&a.source.as_ref().unwrap(), offset, length);
    }

    pub fn join(mut spans: Vec<Span>) -> Span {
        let mut combined = match spans.pop() {
            Some(span) => span,
            None       => return Span::empty(),
        };

        while let Some(span) = spans.pop() {
            combined = Span::combine(&combined, &span)
        }

        return combined;
    }
}

impl<T> Spanned<T> {
    pub fn new(item: T, span: Span) -> Spanned<T> {
        Spanned { item, span }
    }

    fn join(mut spans: Vec<Span>) -> Span {
        let mut combined = match spans.pop() {
            Some(span) => span,
            None => Span::empty()
        };
    
        while let Some(span) = spans.pop() {
            combined = Span::combine(&combined, &span);
        }
    
        combined
    }
    

    pub fn build(spanneds: &Vec<Spanned<T>>) -> Span {
        let spans: Vec<_> = spanneds.iter().map(|s| s.span.clone()).collect();
       
        Spanned::<T>::join(spans)
    }
}
