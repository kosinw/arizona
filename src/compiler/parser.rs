use crate::{
    common::span::*, compiler::ast, compiler::error::*, compiler::lexer::*, compiler::token::*,
};

pub type PResult<T> = Result<T, ParserError>;

struct Parser {
    tokens: Tokens,
    cursor: usize,
    errors: ParserErrors,
}

impl Parser {
    fn new(tokens: Tokens) -> Parser {
        Parser {
            tokens,
            cursor: 0,
            errors: vec![],
        }
    }

    // NOTE: This will EXPLODE if self.tokens is empty
    fn peek(&mut self) -> PResult<Spanned<Token>> {
        match self.tokens.get(self.cursor) {
            Some(token) => Ok(token.clone()),
            None => Err(ParserError::new(
                ParserErrorKind::UnexpectedEOF,
                &self.tokens.last().unwrap().span,
            )),
        }
    }

    fn peek_token(&mut self) -> PResult<Token> {
        let Spanned { item: tok, .. } = self.peek()?;
        Ok(tok)
    }

    fn expect(&mut self, expected: &Token) -> PResult<Span> {
        let Spanned { item: got, span } = self.peek()?;

        if *expected != got {
            return Err(ParserError::new(
                ParserErrorKind::UnexpectedToken {
                    expected: expected.clone(),
                    got: got.clone(),
                },
                &span,
            ));
        }

        self.cursor += 1;

        Ok(span)
    }

    fn end_expr(&mut self) -> PResult<Span> {
        let Spanned { item: got, span } = self.peek()?;

        match got {
            Token::End | Token::Semicolon | Token::Newline | Token::Comma => {
                self.cursor += 1;
                Ok(span)
            }
            _ => Err(ParserError::new(
                ParserErrorKind::ExpectedEndExpr(got),
                &span,
            )),
        }
    }

    fn parse_expr(&mut self) -> PResult<ast::Expr> {
        match self.peek_token()? {
            Token::Let => self.parse_let_expr(),
            Token::IntegerLiteral(_) => self.parse_int_literal_expr(),
            _ => todo!(),
        }
    }

    fn parse_let_expr(&mut self) -> PResult<ast::Expr> {
        let first = self.expect(&Token::Let)?;

        let lhs = self.parse_ident()?;

        let rhs = if self.peek_token()? == Token::Assign {
            self.expect(&Token::Assign)?;

            Some(Box::new(self.parse_expr()?))
        } else {
            None
        };

        let last = self.end_expr()?;

        // TODO: Add types stuff
        Ok(ast::Expr {
            span: Span::combine(&first, &last),
            item: ast::ExprKind::LetExpr(lhs, None, rhs),
        })
    }

    // TODO: This will just not work out in the end
    fn parse_int_literal_expr(&mut self) -> PResult<ast::Expr> {
        let Spanned { item: token, span } = self.peek()?;

        match &token {
            Token::IntegerLiteral(i) => {
                self.cursor += 1;
                Ok(ast::Expr {
                    item: ast::ExprKind::LiteralExpr(ast::Literal::Integer(*i)),
                    span,
                })
            }
            t => Err(ParserError::new(
                ParserErrorKind::ExpectedInteger(t.clone()),
                &span,
            )),
        }
    }

    fn parse_ident(&mut self) -> PResult<Spanned<String>> {
        let Spanned { span, item: tok } = self.peek()?;

        let s = match tok {
            Token::Identifier(s) => s,
            t => {
                return Err(ParserError::new(
                    ParserErrorKind::ExpectedIdentifier(t),
                    &span,
                ))
            }
        };

        self.cursor += 1;

        Ok(Spanned { item: s, span })
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use crate::common::source::*;
    use insta::assert_debug_snapshot;

    #[test]
    fn local_decl() {
        let source = Source::pathless("let i;");
        let tokens = lex(&source).unwrap();

        let mut parser = Parser::new(tokens);
        let let_stmt = parser.parse_expr();

        assert_debug_snapshot!(let_stmt);
    }

    #[test]
    fn local_int_init() {
        let source = Source::pathless("let i = 13 ");
        let tokens = lex(&source).unwrap();

        let mut parser = Parser::new(tokens);
        let let_stmt = parser.parse_expr();

        assert_debug_snapshot!(let_stmt);
    }
}
