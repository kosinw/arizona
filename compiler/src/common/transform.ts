import { SyntaxNode } from '../parser/types';
import {
  Nullable,
  TransformCallback,
  TransformLifetimeCallback,
  TransformMap,
  TransformMapEntry,
} from './types';

// Post-order AST walker
export function transform<T>(map: TransformMap<T>) {
  const walk = (node: SyntaxNode): Nullable<T> => {
    if (node === null) {
      return null;
    }
    const { params } = node;

    let mappingFunction: TransformCallback<T> = (_, __) => {};
    let preCallback: Nullable<TransformLifetimeCallback> = null;
    let postCallback: Nullable<TransformLifetimeCallback> = null;

    if (node.type in map && typeof map[node.type] === 'function') {
      mappingFunction = map[node.type]! as TransformCallback<T>;
    } else if (node.type in map && typeof map[node.type] !== 'function') {
      const entry = map[node.type]! as TransformMapEntry<T>;
      mappingFunction = entry.callback;
      preCallback = entry.pre || null;
      postCallback = entry.post || null;
    }
    if (preCallback) {
      preCallback(node);
    }
    if (params.length === 0) {
      let result = mappingFunction(node, []);
      if (postCallback) {
        postCallback(node);
      }
      return !!result ? result : null;
    } else {
      const transformedParams = params.map((n) => {
        n.parent = node;
        return walk(n);
      });
      let result = mappingFunction(node, transformedParams);
      if (postCallback) {
        postCallback(node);
      }
      return !!result ? result : null;
    }
  };
  return walk;
}
