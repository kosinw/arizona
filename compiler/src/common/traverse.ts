import { SyntaxNode } from '../parser/types';
import { TraverseMap } from './types';

// Pre-order AST walker
export const traverse = (map: TraverseMap) => {
  const walk = (node: SyntaxNode) => {
    if (node === null) {
      return node;
    }
    const { params } = node;
    const mappingFunction = (() => {
      if ('*' in map && typeof map['*'] === 'function') {
        return map['*'];
      }
      if (node.type in map && typeof map[node.type] === 'function') {
        return map[node.type];
      }
      return () => node;
    })();
    if (mappingFunction!.length === 2) {
      mappingFunction!(node, walk);
      return node;
    }
    mappingFunction!(node);
    params.forEach(walk);
    return node;
  };
  return walk;
};
