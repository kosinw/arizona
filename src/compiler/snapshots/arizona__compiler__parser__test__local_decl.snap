---
source: src/compiler/parser.rs
expression: let_stmt

---
Ok(
    Spanned {
        item: LetExpr(
            Spanned {
                item: "i",
                span: Span {
                    offset: 4,
                    length: 1,
                    source: Some(
                        Source {
                            contents: "let i;",
                            path: "<stdin>",
                            line_starts: [
                                0,
                            ],
                        },
                    ),
                },
            },
            None,
            None,
        ),
        span: Span {
            offset: 0,
            length: 6,
            source: Some(
                Source {
                    contents: "let i;",
                    path: "<stdin>",
                    line_starts: [
                        0,
                    ],
                },
            ),
        },
    },
)
